### PR TITLE
RFC: UDP support

### PR DESCRIPTION
**Pull request by grrtrr**
Addresses #215 - "transport: add UDP support?"

This is for review -- please see commit message and the code.

Passages requiring input have been marked by `FIXME:` comments.

If the basic approach is ok, will resubmit with the matching test cases.

I had spent much time looking at alternative implementations; the first one would have added a packet-based `connpacket` connection type to `conn.go`.

Did not like that, since it modified a mangos core file, hence spent time to refine this one, i.e. it is self-contained.

### CODE REVIEW

File: transport/udp/udp.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/go-mangos/mangos"
+)
+
+// Errors produced by this transport prototocol
+var (
```
Comment (gdamore) [inline transport/udp/udp.go]: Just put this in errors.go.

File: transport/udp/udp.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+	ErrMsgTruncated = errors.New("message truncated due to insufficient buffer space")
+)
+
+// connHeader is exchanged during the initial handshake.
+// FIXME: stolen from conn.go -- unify this
```
Comment (gdamore) [inline transport/udp/udp.go]: Agreed, just use conn.go's header.  But lets make sure that the headers are the same.  Alternatively, this could be "udpHeader", which happens to be the same but lets the two evolve separately.

File: transport/udp/udp.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+	Version byte // only zero at present
+	Proto   uint16
+	Rsvd    uint16 // always zero at present
+}
+
+const (
+	// Maximum read queue depth for listening endpoints
+	// FIXME: instead of udpMaxBacklog, use a ListenerOption to set this
+	// FIXME: e.g. OptionUdpListenerBacklog
+	udpMaxBacklog = 100
```
Comment (gdamore) [inline transport/udp/udp.go]: Just OptionBacklog ?  Isn't this analogous to the the listenq depth?

File: transport/udp/udp.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+	// Maximum expected UDP datagram size.
+	// UDPv4 datagrams can be up to 65507 = 2^16-1 - 20 - 8 bytes, due to the 2-byte
+	// IPv4 total length field, a minimum IPv4 header size of 20 bytes, and the 8
+	// byte UDP header length. IPv6 minimum header size is 40 bytes, so it is slightly
+	// less (not considering UDPv6 jumbograms, RFC 2675).
+	// These are theoretical limits, since in practice the maximum datagram size
+	// is determined by the MTU (typically 1500 .. ~9000 bytes).
+	// Hence a buffer size of 64k should suffice for 'regular' networks.
+	udpMaxBuf = 1 << 16
```
Comment (gdamore) [inline transport/udp/udp.go]: This isn't true for UDPv6 messages, which can span a larger message -- up to just a little short of 4GB.

File: transport/udp/udp.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+		var b bytes.Buffer
+		var h = connHeader{S: 'S', P: 'P', Proto: p.proto.Number()}
+
+		if err := binary.Write(&b, binary.BigEndian, &h); err != nil {
+			return err
+		}
+		if _, err := b.Write(m.Header); err != nil {
+			return err
+		}
+		if _, err := b.Write(m.Body); err != nil {
```
Comment (gdamore) [inline transport/udp/udp.go]: Probably we should check for the case of a message that is too big.  We need to return the error case of a message that won't fit into a UDP message to the caller somehow. 

This also underscores our need for statistics...

File: transport/udp/udp.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+// halfPipePeers records remote endpoints that have connected to this Listener
+var halfPipePeers = struct {
+	// which peer connected the listening endpoint
+	byAddr map[string]*udpHalfPipe
+	sync.Mutex
+}{byAddr: make(map[string]*udpHalfPipe)}
+
+// Accept sorts messages into the mailboxes of existing peers, until a new
+// peer connects; causing a new udpHalfPipe to be returned.
```
Comment (gdamore) [inline transport/udp/udp.go]: You've tried to make a pipe/connection oriented interface here.  I think this isn't quite the way I'd go.  UDP is by its very nature disconnected.

I'd instead go with a different implementation, where Listen() does the bind, but Accept() is just a noop to return a suitable reference to the single listening pipe (perhaps wrapped in another object).

I think this would be simpler than the whole full/half pipe thing.

Under this scenario, UDP applications would never see more than a single "pipe" for a given listening address, and
for a given "connect" address they would accept packets from anywhere (or perhaps the option to discard from "unconnected" peers could be a socket option?)

I want to think on this a bit.

File: transport/udp/udp.go [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+	if addr, err = mangos.StripScheme(t, addr); err != nil {
+		return nil, err
+	}
+	if e.addr, err = net.ResolveUDPAddr("udp", addr); err != nil {
+		return nil, err
+	}
+
+	if val, err := sock.GetOption(mangos.OptionMaxRecvSize); err == nil {
+		if maxrx, ok := val.(int); ok && maxrx < udpMaxBuf {
+			return nil, fmt.Errorf("MaxRecvSize(%d) to small for udp - need %d", maxrx, udpMaxBuf)
```
Comment (gdamore) [inline transport/udp/udp.go]: Huh?  This should be payload size, not udpMaxBuf.  I think this is wrong.

Comment (gdamore) [inline transport/udp/udp.go]: Its not entirely clear to me that we need to have a maximum option here, at all, as UDP imposes an inherent limit.  But we can.

We could also allow message reassembly, but that would require a change to the RFC.  I think users who want larger messages shouldn't use UDP, or should use a higher level framing protocol instead.

File: transport/udp/udp.go (hunk 4 above)
Comment (grrtrr) [inline transport/udp/udp.go]: So are you saying you would like to support UDPv6 jumbograms (RFC 2675) with the IPv6 hop-by-hop jumbo payload option? If this is not the case, it falls back to the 16-it UDP header, even on IPv6.

File: transport/udp/udp.go (hunk 7 above)
Comment (grrtrr) [inline transport/udp/udp.go]: I am open to suggestions here. At present, it is just making sure that the user does not expect the socket to receive data larger than the largest-possibly buffer size (which was set for the UDPv4 and the default UDPv6 case).

File: transport/udp/udp.go (hunk 4 above)
Comment (gdamore) [inline transport/udp/udp.go]: Yes, we should be able to support UDPv6 jumbograms.  I don't think the application needs to be aware of this, but I've not really looked too hard at it.

Comment (gdamore) [inline transport/udp/udp.go]: Reading the RFC, it seems that this is restricted only to specific links; so for most people it won't be available.  But one can imagine e.g. shared memory based networks (RDMA, etc.) where the link MTU can be quite large.  For example, we had a shared memory network in the Sun E10K between domains that made use of shared memory between the two domains.  That could have used a very large MTU, say 1MB, without concerns about checksum etc. because it was simply local memory to both nodes.  One can imagine a similar situation with paravirualized networks, although I believe today they all have "smaller" jumbo MTUs.

File: transport/udp/udp.go (hunk 7 above)
Comment (gdamore) [inline transport/udp/udp.go]: But that doesn't seem to be what the code does, if I'm reading it.  It seems like you don't let the user set a smaller maximum receive size.  I can see perfectly valid reasons to limit the max recv size even smaller than the maximum UDP message size.  Especially this is true for jumbo UDP.

Perhaps you mean the sense of the < sign to be reversed?

File: transport/udp/udp.go (hunk 4 above)
Comment (grrtrr) [inline transport/udp/udp.go]: Have you had time to ponder this, and how would you like to proceed?

I am understanding your comment to mean that there may be a different way of integrating UDP support. The patch submitted for review presents one possible alternative I had explored until it worked reliably (in the REQ/REP example).

I am not claiming that this is the best possible solution. Rather I think that there is a deeper question: since the interface is already built around the `Listen`, `Dial` and `Accept` verbs, any integration of UDP will have relative pros and cons. 

There were good engineering reasons in the Berkely sockets API to use different verbs for record-/packet-oriented and for byte-stream oriented sockets; which still seem to hold today.
An attempt to cross the boundaries to me seems a bit like putting a triangular or hexagonal peg into a round hole.

In conclusion, I believe it might be better to approach the integration from a higher level, considering a packet-oriented interface for go-mangos.

But then I ask myself, would the change be worth it? Perhaps keep the interface as it is and keep it with the transports it already provides. One big advantage of that is that it is consistent.

Your thoughts?

**Comment by gdamore** [comment]
I've been thinking about this more lately, and I'm pretty convinced that in order for REQ/REP to work with Devices, we need to change the UDP RFC to include some back trace information.  This means that we need to keep state about "logical connections", even though UDP is connectionless. 

I have a similar problem with another transport I'm working on, that is also connectionless, but needs this ability.

**Comment by grrtrr** [comment]
Yes, I had quite the same problem: endpoint A talking to B, and C talking to B are different. Even in the [QUIC document](https://docs.google.com/document/d/1RNHkx_VvKWyWg6Lr8SZ-saqsQx7rFV-ev2jRFUoVD34/edit#heading=h.rblq0wj4ce5h) they are talking about 'UDP connections'. 

Let me know if I can be of any help, with regard to ideas, or trying a different implementation. For us, the UDP alternative is still attractive due to the reduced time it takes for one message transaction.

**Comment by gdamore** [comment]
You should read the ZeroTier RFC in the nanomsg rfc repository.  This gives an idea of what I'm thinking about -- I'll be updating the UDP RFC to support a similar strategy soon.

**Comment by grrtrr** [comment]
This is exciting news! Should this PR be closed, is there a better place to communicate (e.g. #215)?

I am looking forward to how this evolves. Two short questions:
- As there now is a notion of _logical connections_, is reusing the `Pipe` interface now a better fit?
- The ZeroTier has quite a few message types. Do you intend to keep the UDP communication fairly simple (in contrast)? I have seen in [DCCP](https://www.ietf.org/rfc/rfc4340.txt) how many different message types (10) existed in contrast to a single UDP datagram.

Thank you for keeping this up.

**Comment by gdamore** [comment]
Probably yes, this PR should be closed, but leave a github issue open for the request.  I need to update the RFC and will begin doing so soon.  I anticipate writing this in nanomsg ahead of mangos, owing to the fact that the nanomsg work is funded.

I don't plan to support all the message types that ZeroTier has, but will probably have some subset of them necessary for maintaining logical connections.

**Comment by grrtrr** [comment]
Closing this, leaving #215 open. Will keep looking at nanomsg updates. With regard to making UDP stateful, in DCCP I have seen this becoming quite complicated very quick. Perhaps it is possible to build upon existing efforts, as e.g. QUIC?
